File: backend/ai/app/service/conversation.py
"""会话历史的读写：Redis 缓存 + PostgreSQL 持久化（cache-aside）。

读：先查 Redis，没有再查 PostgreSQL 并回填缓存。
写：先写 PostgreSQL（唯一真相源），成功后再刷新 Redis。

入参 user_id 统一是字符串（来自网关注入的 X-Kada-User-ID 请求头），
落库查询时再转成 int（数据库列是整数）。
"""

import json
import uuid

from sqlmodel import select
from sqlmodel.ext.asyncio.session import AsyncSession

# 必须用 SQLModel 自带的 AsyncSession（有 .exec()，和 select() 搭配）；
# 误用 sqlalchemy.ext.asyncio.AsyncSession 会报没有 exec 方法。
from app.config import settings
from app.service.db import Conversation, Message, engine
from app.service.redis_client import get_redis


def _cache_key(user_id: str, conversation_id: str) -> str:
    # key 里带 user_id，不同用户的会话天然隔离。
    return f"ai:conv:{user_id}:{conversation_id}"


async def load_messages(user_id: str, conversation_id: str) -> list:
    """取某会话的历史；任何一层都没有（新会话）时返回空列表。"""
    key = _cache_key(user_id, conversation_id)
    cid = uuid.UUID(conversation_id)

    # 1) 先查缓存，命中则一条数据库查询都不用跑。
    redis = await get_redis()
    if redis is not None:
        data = await redis.get(key)
        if data:
            return json.loads(data)

    # 2) 缓存未命中，查 PostgreSQL。会话归属条件直接写进 WHERE：
    #    不属于该用户的会话根本查不出来（所有权在查询层保证）。
    async with AsyncSession(engine) as session:
        conversation = (
            await session.exec(
                select(Conversation).where(
                    Conversation.id == cid,
                    Conversation.user_id == int(user_id),
                )
            )
        ).first()
        if conversation is None:
            return []

        rows = (
            await session.exec(
                select(Message)
                .where(Message.conversation_id == cid)
                .order_by(Message.created_at)
            )
        ).all()

    # 只保留最近 MAX_MESSAGES 条，和缓存口径保持一致。
    messages = [
        {"role": m.role, "content": m.content} for m in rows
    ][-settings.MAX_MESSAGES :]

    # 3) 回填缓存，下次读取直接命中。
    if messages and redis is not None:
        await redis.set(
            key,
            json.dumps(messages, ensure_ascii=False),
            ex=settings.CACHE_TTL_SECONDS,
        )
    return messages


async def new_conversation(user_id: str) -> str:
    """新建一条会话记录，返回会话 id。"""
    async with AsyncSession(engine) as session:
        conversation = Conversation(user_id=int(user_id))
        session.add(conversation)
        await session.commit()
        await session.refresh(conversation)
        return str(conversation.id)


async def append_message(
    user_id: str, conversation_id: str, role: str, content: str
) -> None:
    """追加一条消息。先写库再刷缓存；会话不属于该用户时直接忽略。"""
    cid = uuid.UUID(conversation_id)

    async with AsyncSession(engine) as session:
        conversation = (
            await session.exec(
                select(Conversation).where(
                    Conversation.id == cid,
                    Conversation.user_id == int(user_id),
                )
            )
        ).first()
        if conversation is None:
            return

        session.add(Message(conversation_id=cid, role=role, content=content))
        await session.commit()

    # 缓存里有就就地追加（省一次全量回查）；没有则不主动写，
    # 留给下次读取时回填，避免在写路径上重复查库。
    redis = await get_redis()
    if redis is not None:
        key = _cache_key(user_id, conversation_id)
        data = await redis.get(key)
        if data:
            messages = json.loads(data)
            messages.append({"role": role, "content": content})
            await redis.set(
                key,
                json.dumps(messages[-settings.MAX_MESSAGES :], ensure_ascii=False),
                ex=settings.CACHE_TTL_SECONDS,
            )


async def list_conversations(user_id: str) -> list:
    """列出该用户的会话（最新在前），供单会话模式取"当前会话"。"""
    async with AsyncSession(engine) as session:
        result = await session.exec(
            select(Conversation)
            .where(Conversation.user_id == int(user_id))
            .order_by(Conversation.created_at.desc())
        )
        conversations = result.all()
    return [
        {
            "conversation_id": str(c.id),
            "created_at": c.created_at.isoformat(),
        }
        for c in conversations
    ]


async def delete_conversation(user_id: str, conversation_id: str) -> bool:
    """删除单个会话（消息靠数据库 ON DELETE CASCADE 连带删除）并清缓存。"""
    cid = uuid.UUID(conversation_id)
    async with AsyncSession(engine) as session:
        conversation = (
            await session.exec(
                select(Conversation).where(
                    Conversation.id == cid,
                    Conversation.user_id == int(user_id),
                )
            )
        ).first()
        if conversation is None:
            return False
        await session.delete(conversation)
        await session.commit()

    redis = await get_redis()
    if redis is not None:
        await redis.delete(_cache_key(user_id, conversation_id))
    return True


async def get_current_conversation(user_id: str) -> dict:
    """单会话模式：返回该用户最新一个会话及其消息；没有则返回空。"""
    conversations = await list_conversations(user_id)
    if not conversations:
        return {"conversation_id": None, "messages": []}
    cid = conversations[0]["conversation_id"]
    messages = await load_messages(user_id, cid)
    return {"conversation_id": cid, "messages": messages}


async def restart_conversation(user_id: str) -> str:
    """重新开始：删掉该用户所有旧会话（含缓存），返回全新空会话 id。"""
    conversations = await list_conversations(user_id)
    for conversation in conversations:
        await delete_conversation(user_id, conversation["conversation_id"])
    return await new_conversation(user_id)

File: backend/internal/assistant/conversation.go
package assistant

import (
	"context"
	"errors"
	"fmt"

	"github.com/google/uuid"
	"gorm.io/gorm"

	"github.com/chun/kada-backend/internal/domain/entity"
)

// historyLimit is how many turns the model is given. It is the same 50 the Python service used
// (settings.MAX_MESSAGES): a context-window decision, not an implementation detail of the port.
const historyLimit = 50

// ChatMessage is one turn of a conversation, in the shape the model and the frontend both read.
type ChatMessage struct {
	Role    string `json:"role"`
	Content string `json:"content"`
}

// Conversation is a session as the frontend sees it. ConversationID is null for a user who has never
// chatted, which is what the AI page renders as an empty history.
type Conversation struct {
	ConversationID *string       `json:"conversation_id"`
	Messages       []ChatMessage `json:"messages"`
}

// Conversations stores chat history in PostgreSQL.
//
// It is the only copy. The Python service kept a Redis cache in front of the same data; that is gone on
// purpose - the reads are one indexed SELECT per question, and a cache that can disagree with the
// database is one more thing to reason about for no measurable gain at this size.
type Conversations struct {
	db *gorm.DB
}

// NewConversations builds the store on an open database handle.
func NewConversations(db *gorm.DB) *Conversations {
	return &Conversations{db: db}
}

// Start creates an empty conversation for the user and returns its id.
func (c *Conversations) Start(ctx context.Context, userID int64) (string, error) {
	conversation := entity.AIConversation{ID: uuid.New(), UserID: userID}
	if err := c.db.WithContext(ctx).Create(&conversation).Error; err != nil {
		return "", fmt.Errorf("failed to start a conversation: %w", err)
	}
	return conversation.ID.String(), nil
}

// Load returns the recent turns of a conversation, oldest first.
//
// A conversation that does not exist, or belongs to somebody else, yields no turns and no error: the
// ownership condition is in the query, so the rows simply do not match, and the caller treats that as a
// new conversation. An id that is not a UUID is the same thing - it cannot be a conversation this user
// owns.
func (c *Conversations) Load(ctx context.Context, userID int64, conversationID string) ([]ChatMessage, error) {
	id, err := uuid.Parse(conversationID)
	if err != nil {
		return nil, nil //nolint:nilerr // an unparsable id is not an error, it is not this user's conversation
	}

	var rows []entity.AIMessage
	if err := c.historyQuery(c.db.WithContext(ctx), userID, id).Find(&rows).Error; err != nil {
		return nil, fmt.Errorf("failed to load the conversation: %w", err)
	}
	return toMessages(rows), nil
}

// Append stores one turn. A conversation that does not belong to the user is ignored rather than
// reported: the only way to reach that branch is a forged or stale id from the client, and the caller has
// nothing useful to do with the error.
func (c *Conversations) Append(ctx context.Context, userID int64, conversationID, role, content string) error {
	id, err := uuid.Parse(conversationID)
	if err != nil {
		return nil //nolint:nilerr // see Load: the id is not a conversation, so there is nothing to append to
	}

	owned, err := c.owned(ctx, userID, id)
	if err != nil {
		return err
	}
	if !owned {
		return nil
	}

	message := entity.AIMessage{
		ID:             uuid.New(),
		ConversationID: id,
		Role:           role,
		Content:        content,
	}
	if err := c.db.WithContext(ctx).Create(&message).Error; err != nil {
		return fmt.Errorf("failed to store the message: %w", err)
	}
	return nil
}

// Current returns the user's latest conversation with its turns, or an empty conversation.
func (c *Conversations) Current(ctx context.Context, userID int64) (Conversation, error) {
	var latest entity.AIConversation
	err := c.latestQuery(c.db.WithContext(ctx), userID).First(&latest).Error
	if errors.Is(err, gorm.ErrRecordNotFound) {
		return Conversation{Messages: []ChatMessage{}}, nil
	}
	if err != nil {
		return Conversation{}, fmt.Errorf("failed to load the latest conversation: %w", err)
	}

	messages, err := c.Load(ctx, userID, latest.ID.String())
	if err != nil {
		return Conversation{}, err
	}
	id := latest.ID.String()
	return Conversation{ConversationID: &id, Messages: messages}, nil
}

// Restart drops every conversation the user has and starts a new one.
//
// The messages go with them through ON DELETE CASCADE, which is declared on the model and therefore in the
// database: a delete here cannot leave turns behind.
func (c *Conversations) Restart(ctx context.Context, userID int64) (string, error) {
	err := c.db.WithContext(ctx).Where("user_id = ?", userID).Delete(&entity.AIConversation{}).Error
	if err != nil {
		return "", fmt.Errorf("failed to drop the previous conversations: %w", err)
	}
	return c.Start(ctx, userID)
}

// owned reports whether the conversation belongs to the user.
func (c *Conversations) owned(ctx context.Context, userID int64, id uuid.UUID) (bool, error) {
	var count int64
	err := c.db.WithContext(ctx).
		Model(&entity.AIConversation{}).
		Where("id = ? AND user_id = ?", id, userID).
		Count(&count).Error
	if err != nil {
		return false, fmt.Errorf("failed to check the conversation owner: %w", err)
	}
	return count > 0, nil
}

// historyQuery selects a conversation's turns, newest first.
//
// The join is what enforces ownership: a turn is only reached through a conversation that belongs to this
// user, and the limit therefore always trims the *old* end of that user's own history.
func (c *Conversations) historyQuery(query *gorm.DB, userID int64, id uuid.UUID) *gorm.DB {
	return query.
		Model(&entity.AIMessage{}).
		Select("ai_messages.*").
		Joins("JOIN ai_conversations ON ai_conversations.id = ai_messages.conversation_id").
		Where("ai_messages.conversation_id = ? AND ai_conversations.user_id = ?", id, userID).
		Order("ai_messages.created_at DESC").
		Limit(historyLimit)
}

// latestQuery selects the user's most recent conversation.
func (c *Conversations) latestQuery(query *gorm.DB, userID int64) *gorm.DB {
	return query.
		Model(&entity.AIConversation{}).
		Where("user_id = ?", userID).
		Order("created_at DESC")
}

// toMessages turns stored rows into turns in reading order.
//
// The query takes the newest rows (DESC) so that the limit keeps the recent end of a long conversation;
// the model has to read them oldest first, so they are reversed here.
func toMessages(rows []entity.AIMessage) []ChatMessage {
	messages := make([]ChatMessage, 0, len(rows))
	for i := len(rows) - 1; i >= 0; i-- {
		messages = append(messages, ChatMessage{Role: rows[i].Role, Content: rows[i].Content})
	}
	return messages
}
